test/java/us/klette/constantstring/OperationsTest.java

```diff
@@ -0,0 +1,156 @@
+
+/**
+ * Test cases for basic operations.
+ *
+ * @author Kristian Klette (klette@klette.us)
+ */
+public final class OperationsTest {
+
+    @Test
+    public void create() throws Exception {
+        CString foobar = CString.create("foobar");
+
+        assertThat(foobar.toString()).isEqualTo("foobar");
+    }
+
+    /**
+     * Verify that CString can store and evaluate basic strings.
+     *
+     * @throws Exception e
+     */
+    @Test
+    public void evalShouldReturnBasicValue() throws Exception {
+        final String string = "basic";
+        final CString cString = CString.create(string);
+        final String eval = cString.toString();
+        assertThat(eval).isEqualTo(string);
+    }
+
+    /**
+     * Verify that we can substring a value once.
+     *
+     * @throws Exception e
+     */
+    @Test
+    public void shouldBeAbleToSubstring() throws Exception {
+        final String string = "substr";
+        final int index = 3;
+        final CString initial = CString.create(string);
+        final CString substring = initial.substring(index);
+        final String evaluated = substring.toString();
+        assertThat(evaluated).isEqualTo(string.substring(index));
+    }
+
+    /**
+     * Verify that substring operations with an index higher than
+     * the length of the underlying CString, then an empty string
+     * should be returned.
+     */
+    @Test
+    public void substringWithIndexOutsideBoundsShouldBeEmptyString() {
+        final String foobar = "foobar";
+        final String substr = CString
+                .create(foobar)
+                .substring(foobar.length())
+                .toString();
+        assertThat(substr).isEqualTo("");
+    }
+
+    /**
+     * Verify that substrings of substrings are working.
+     *
+     * @throws Exception e
+     */
+    @Test
+    public void substringOfSubstring() throws Exception {
+        final CString initial = CString.create("abc");
+        final CString first = initial.substring(1);
+        final CString second = first.substring(1);
+        final String evaluated = second.toString();
+        assertThat(evaluated).isEqualTo("c");
+    }
+
+    @Test
+    public void concatTest() throws Exception {
+        final CString foo = CString.create("foo");
+        final CString bar = CString.create("bar");
+        final CString foobar = foo.concat(bar);
+
+        assertThat(foobar.toString()).isEqualTo("foobar");
+    }
+    @Test
+    public void concatStringTest() throws Exception {
+        final CString foo = CString.create("foo");
+        final CString foobar = foo.concat("bar");
+
+        assertThat(foobar.toString()).isEqualTo("foobar");
+    }
+
+    @Test
+    public void delete() throws Exception {
+        CString foobar = CString.create("foobar");
+        CString bar = foobar.delete(0, 3);
+
+        assertThat(bar.toString()).isEqualTo("bar");
+    }
+
+    @Test
+    public void substring() throws Exception {
```

Comment:
haven't I seen this test before?? And aren't there multiple dufferent cases of substrings? Do we test all of them here?